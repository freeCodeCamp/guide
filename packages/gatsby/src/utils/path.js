const path = require(`path`)
const os = require(`os`)

export function joinPath(...paths) {
  const joinedPath = path.join(...paths)
  if (os.platform() === `win32`) {
    return joinedPath.replace(/\\/g, `\\\\`)
  } else {
    return joinedPath
  }
}
