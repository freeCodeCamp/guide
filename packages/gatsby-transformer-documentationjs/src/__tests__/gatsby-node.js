import groupBy from "lodash/groupBy"
import path from "path"
import gatsbyNode from "../gatsby-node"

describe(`transformer-react-doc-gen: onCreateNode`, () => {
  let boundActionCreators, node, createdNodes, updatedNodes
  let run = (node = node, opts = {}) =>
    gatsbyNode.onCreateNode(
      {
        node,
        boundActionCreators,
      },
      opts
    )

  beforeEach(() => {
    createdNodes = []
    updatedNodes = []
    node = {
      id: `node_1`,
      children: [],
      absolutePath: path.join(__dirname, `fixtures`, `code.js`),
      internal: {
        mediaType: `application/javascript`,
        type: `File`,
      },
    }
    boundActionCreators = {
      createNode: jest.fn(n => createdNodes.push(n)),
      createParentChildLink: jest.fn(n => updatedNodes.push(n)),
    }
  })

  it(`should extract out a description, params, and examples`, async () => {
    await run(node)
    expect(createdNodes).toMatchSnapshot()
  })

  it(`should only process javascript File nodes`, async () => {
    let result
    result = await run({ internal: { mediaType: `text/x-foo` } })
    expect(result).toBeNull()

    result = await run({ internal: { mediaType: `application/javascript` } })
    expect(result).toBeNull()

    result = await run({
      id: `test`,
      children: [],
      absolutePath: path.join(__dirname, `fixtures`, `code.js`),
      internal: { mediaType: `application/javascript`, type: `File` },
    })
    expect(createdNodes.length).toBeGreaterThan(0)
  })

  it(`should extract create description nodes with markdown types`, async () => {
    await run(node)
    let types = groupBy(createdNodes, `internal.type`)
    expect(
      types.DocumentationJSComponentDescription.every(
        d => d.internal.mediaType === `text/x-markdown`
      )
    ).toBe(true)
  })
})
