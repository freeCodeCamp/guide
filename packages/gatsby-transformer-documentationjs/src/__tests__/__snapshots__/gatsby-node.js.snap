// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformer-react-doc-gen: onCreateNode should extract out a description, params, and examples 1`] = `
Array [
  Object {
    "children": Array [],
    "id": "documentationJS node_1 comment #0--DocumentationJSComponentDescription--comment.description",
    "internal": Object {
      "content": "A pretty cool jsdoc example
",
      "contentDigest": "1effd106a9028bac69276a250a0f96f5",
      "mediaType": "text/x-markdown",
      "type": "DocumentationJSComponentDescription",
    },
    "parent": "node_1",
  },
  Object {
    "children": Array [],
    "id": "documentationJS node_1 comment #0--DocumentationJSComponentDescription--apple",
    "internal": Object {
      "content": "A nice crispy apple
",
      "contentDigest": "bdc56b714e173673c86ecf5f556c5fd1",
      "mediaType": "text/x-markdown",
      "type": "DocumentationJSComponentDescription",
    },
    "parent": "node_1",
  },
  Object {
    "children": Array [],
    "commentNumber": 0,
    "description___NODE": "documentationJS node_1 comment #0--DocumentationJSComponentDescription--comment.description",
    "examples": Array [
      Object {
        "highlighted": "<span class=\\"token keyword\\">const</span> apple <span class=\\"token operator\\">=</span> <span class=\\"token function\\">require</span><span class=\\"token punctuation\\">(</span><span class=\\"token string\\">'apple'</span><span class=\\"token punctuation\\">)</span>
<span class=\\"token function\\">apple</span><span class=\\"token punctuation\\">(</span><span class=\\"token punctuation\\">)</span>",
        "raw": "const apple = require('apple')
apple()",
      },
    ],
    "id": "documentationJS node_1 comment #0",
    "internal": Object {
      "content": "{
    \\"kind\\": \\"constant\\",
    \\"name\\": \\"apple\\",
    \\"params\\": [
        {
            \\"title\\": \\"param\\",
            \\"name\\": \\"apple\\",
            \\"type\\": {
                \\"type\\": \\"NameExpression\\",
                \\"name\\": \\"string\\"
            },
            \\"description___NODE\\": \\"documentationJS node_1 comment #0--DocumentationJSComponentDescription--apple\\"
        }
    ],
    \\"returns\\": [],
    \\"examples\\": [
        {
            \\"raw\\": \\"const apple = require('apple')\\\\napple()\\",
            \\"highlighted\\": \\"<span class=\\\\\\"token keyword\\\\\\">const</span> apple <span class=\\\\\\"token operator\\\\\\">=</span> <span class=\\\\\\"token function\\\\\\">require</span><span class=\\\\\\"token punctuation\\\\\\">(</span><span class=\\\\\\"token string\\\\\\">'apple'</span><span class=\\\\\\"token punctuation\\\\\\">)</span>\\\\n<span class=\\\\\\"token function\\\\\\">apple</span><span class=\\\\\\"token punctuation\\\\\\">(</span><span class=\\\\\\"token punctuation\\\\\\">)</span>\\"
        }
    ],
    \\"description___NODE\\": \\"documentationJS node_1 comment #0--DocumentationJSComponentDescription--comment.description\\"
}",
      "contentDigest": "b635e85506044a6cdcebbcd7fda00b7d",
      "mediaType": "text/x-javascript-metadata",
      "type": "DocumentationJs",
    },
    "kind": "constant",
    "name": "apple",
    "params": Array [
      Object {
        "description___NODE": "documentationJS node_1 comment #0--DocumentationJSComponentDescription--apple",
        "name": "apple",
        "title": "param",
        "type": Object {
          "name": "string",
          "type": "NameExpression",
        },
      },
    ],
    "parent": "node_1",
    "returns": Array [],
  },
]
`;
